### PR TITLE
feat: Add support for additional volumes in launch templates and launch configurations

### PR DESCRIPTION
**Pull request by jaimehrubiks**
# PR o'clock

## Description

Now the user can specify additional volumes for its instances. This is
specially useful when the cluster is going to be used for specific
applications such as storage.

Example 1 (Just required fields)

    additional_volumes = [{
        block_device_name = "/dev/sdb",
        volume_size = 200,
        delete_on_termination = false
    }]

Example 2 (Multiple additional volumes & All features)

    additional_volumes = [{
        block_device_name = "/dev/sdb",
        volume_size = 200,
        volume_type = "io1"
        iops = 100
        encrypted = true
        kms_key_id = "arn:aws:kms:us-east-1:89349:key/a4-ds-34"
        delete_on_termination = false
    },{
        block_device_name = "/dev/sdc",
        volume_size = 500,
        delete_on_termination = false
    }]

Relevant Issues:

https://github.com/terraform-aws-modules/terraform-aws-eks/issues/687
https://github.com/terraform-aws-modules/terraform-aws-eks/pull/572

Note: I only implemented this feature on launch template workers

### Checklist

- [ ] Change added to CHANGELOG.md. All changes must be added and breaking changes and highlighted
- [ ] CI tests are passing
- [ ] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by jaimehrubiks** [comment]
Why is the code format failing on the local.tf?

**Comment by js-timbirkett** [comment]
Is there a valid use case for extra volumes mounted to worker nodes? Can the same be done with managed nodes?

My thinking is that Kubernetes nodes are autoscaled and would be recycled / scaled up and down by the Cluster Autoscaler. EBS volumes are generally destroyed when an instance is terminated.

I'd think that stateful workloads in need of block storage such as Kafka, MySQL or Ceph would use PersistentVolume's and PersistentVolumeClaims along with an appropriate CSI driver to create, mount and handle lifecycle of volumes?

I might be wrong and may have missed some valuable context somewhere. I'm also not an expert in that field so always appreciate these learning opportunities 😄

**Comment by jaimehrubiks** [comment]
First, I am also learning, not an expert at all. For most stateful applications, as you mention, the CSI way to claim storage is the way to request storage. However, for storage itself, say, Ceph, I found that it might be easier to manage and control to manually specify or add devices, and have a list of specific nodes and their devices. In my case, I was using rook.io, and to make it similar to on-prem, this was the only way. Of course, I would love to see other people's opinion on this matter.

**Comment by snstanton** [comment]
With the new AWS bottlerocket os, the AMI for the worker nodes automatically mounts a secondary volume with a default size of 20GB to act as the containerd cache for the node.  In order to change the size of this volume, I need the ability to override this default size to something larger.   I literally just wrote something very similar to this patch, so it would be great to get the merged in.

**Comment by snstanton** [comment]
It looks like the lint issues are a missing white space at `workers_launch_template.tf:355` and too much white space before the comment at `local.tf:77`.  If you run `terraform fmt` on the files, it will fix the lint issues for you.

**Comment by jaimehrubiks** [comment]
I messed up with my git a little bit

**Comment by snstanton** [comment]
Yes, in fact I branched and added this PR and it works perfectly.  Here's what I have in my config:
```
      additional_volumes     = [{ block_device_name = "/dev/xvdb", volume_size = 50 }]
```

**Comment by snstanton** [comment]
You can fix the Semantic Pull Request failure by changing the PR title to something like:

`feat: add support for additional volumes in launch templates`

File: workers_launch_template.tf [hunk 1]
```diff
@@ -351,6 +351,43 @@ resource "aws_launch_template" "workers_launch_template" {
     }
   }
 
+  dynamic "block_device_mappings" {
+    for_each = lookup(var.worker_groups_launch_template[count.index], "additional_volumes", lookup(local.workers_group_defaults, "additional_volumes", []))
```
Comment (barryib) [inline workers_launch_template.tf]:
```suggestion
    for_each = lookup(var.worker_groups_launch_template[count.index], "additional_volumes", local.workers_group_defaults["additional_volumes"])
```

for consistency and `local.workers_group_defaults["additional_volumes"]` is already set by default to `[]`

File: workers_launch_template.tf [hunk 2]
```diff
@@ -351,6 +351,43 @@ resource "aws_launch_template" "workers_launch_template" {
...
+    content {
+      device_name = block_device_mappings.value.block_device_name
+
+      ebs {
```
Comment (barryib) [inline workers_launch_template.tf]: Does it make sense to rename `local.additional_volumes` into `local. additional_ebs_volumes` ? It sounds that we are only dealing with EBS and not ephemeral storage.

@jaimehrubiks @js-timbirkett @max-rocket-internet any thoughts ?

Comment (jaimehrubiks) [reply workers_launch_template.tf]: I don't really have an opinion in this matter, both alternatives look good for me.

**Comment by jaimehrubiks** [comment]
Will take a look at launch configuration tomorrow, to see if it is similar.

**Comment by jaimehrubiks** [comment]
All changes applied. I did not test this feature on working groups (launch configuration)

File: workers_launch_template.tf (hunk 2 above)
Comment (jaimehrubiks) [reply workers_launch_template.tf]: accepted as makes sense

**Comment by barryib** [comment]
> All changes applied. I did not test this feature on working groups (launch configuration)

Nice @jaimehrubiks. Thanks for your contribution.

**Pull request merged**